### DISCUSSION TITLE
Cannot connect to endpoint to refresh aws-auth ConfigMap

### DISCUSSION THREAD

**Discussion by BrianMusson** [gruntwork-io/knowledge-base#333] [Help] [answered]
Hello, I recently started seeing this error message. I checked my kubectl commands (setup with kubergrunt eks) and I can reach the cluster without issue. My config map looks ok. Where is it getting the `http://localhost` host setting from? I am using version `v0.84.3`, and for what it's worth this issue cropped up out of the blue while troubleshooting something else.  Thanks!

```
module.eks_k8s_role_mapping.kubernetes_config_map.eks_to_k8s_role_mapping: Refreshing state... [id=kube-system/aws-auth]
╷
│ Error: Get "http://localhost/api/v1/namespaces/kube-system/configmaps/aws-auth": dial tcp [::1]:80: connect: connection refused
│
│   with module.eks_k8s_role_mapping.kubernetes_config_map.eks_to_k8s_role_mapping,
│   on .terraform/modules/eks_k8s_role_mapping/modules/eks-k8s-role-mapping/main.tf line 33, in resource "kubernetes_config_map" "eks_to_k8s_role_mapping":
│   33: resource "kubernetes_config_map" "eks_to_k8s_role_mapping" {
│
╵
Releasing state lock. This may take a few moments...
ERRO[0116] 1 error occurred:
	* exit status 1
```

**Comment by autero1** [comment, marked as answer]
This is most likely due to missing / invalid `kubernetes` provider configuration. If you add one - either via `generate` block, `include` or explicit provider configuration, the `kubernetes` provider has correct configuration for API endpoints and auth. 
Without explicit provider configuration, Terraform defaults to localhost and that's why you're seeing the  `http://localhost/api/v1/...` in the output.

The provider generator template can be found here: https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/examples/for-production/infrastructure-live/provider_k8s_helm_for_eks.template.hcl

**Reply by BrianMusson** [reply]
Isn't that setup here: https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/master/modules/services/eks-cluster/main.tf#L41

**Reply by autero1** [reply]
Ahh, you're right! We had a similar case just recently, where the the provider was completely missing for a new module. In your case, I am assuming it's the `eks-cluster` and the embedded `module "eks_k8s_role_mapping"` you're having issues with? 

The issue still seems related to problems with the Kubernetes provider, as it defaults to localhost Kube API. Have you tried clearing all cached modules and providers, reinitializing the terraform config?

**Reply by BrianMusson** [reply]
Yes, thats exactly right.

This is happening in our pipelines as well as on my laptop. In both instances, I have removed the cache dirs.

**Reply by BrianMusson** [reply]
Any ideas what else I can check?

**Reply by rhoboat** [reply]
Can you check your `~/.kube/config` for anything amiss? Maybe something within the `- clusters` list, `cluster.server` URI?

**Reply by rhoboat** [reply]
I have faint confidence in that, since you are already configured correctly and can access the cluster with `kubectl`. 🤔

**Reply by rhoboat** [reply]
Here's some evidence this might be an issue with the provider itself. https://github.com/hashicorp/terraform-provider-kubernetes/issues/546 There are some workarounds people have tried, can you see if this helps?

**Reply by yorinasub17** [reply]
I remember running into this before, and at that time, it was caused by a change in the EKS cluster endpoint due to a cluster recreation and an unclean terraform state with the data sources used to configure the provider. I was able to resolve it then when I ran:

```
terragrunt apply -refresh=false -target data.aws_eks_cluster.cluster
```

Does that work? Also, are you using the `exec` plugin or the data source? Another cause I've seen is when the token is expired. This happens when you are using the data source based auth method with newer terraform versions (0.13+ I think). Before 0.13, terraform always refreshed data sources but starting with 0.13, terraform started tracking data sources in the terraform state and selectively refreshes it only when the inputs change.

If you are using the data source, you can force a refresh by removing the data source from the state and applying it with a `target` call like above.

**Reply by BrianMusson** [reply]
When I attempt your method using the targeting apply, TF wants to recreate the cluster because it is tainted. After removing the data source, should I remove ```module.eks_cluster.aws_eks_cluster.eks``` from the state file and re-import the cluster? 
```
eks-cluster]$ terragrunt state mv data.aws_eks_cluster.cluster data.aws_eks_cluster.cluster_backup                                                            
Acquiring state lock. This may take a few moments...
Move "data.aws_eks_cluster.cluster" to "data.aws_eks_cluster.cluster_backup"
Successfully moved 1 object(s).
Releasing state lock. This may take a few moments...



eks-cluster]$ terragrunt apply -refresh=false -target data.aws_eks_cluster.cluster                                                                            
Acquiring state lock. This may take a few moments...

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement
 <= read (data resources)

Terraform will perform the following actions:

  # data.aws_eks_cluster.cluster will be read during apply
  # (config refers to values not yet known)
 <= data "aws_eks_cluster" "cluster"  {
      + arn                       = (known after apply)
      + certificate_authority     = (known after apply)
      + created_at                = (known after apply)
      + enabled_cluster_log_types = (known after apply)
      + endpoint                  = (known after apply)
      + id                        = (known after apply)
      + identity                  = (known after apply)
      + kubernetes_network_config = (known after apply)
      + name                      = (known after apply)
      + platform_version          = (known after apply)
      + role_arn                  = (known after apply)
      + status                    = (known after apply)
      + tags                      = (known after apply)
      + version                   = (known after apply)
      + vpc_config                = (known after apply)
    }

  # module.eks_cluster.aws_eks_cluster.eks is tainted, so must be replaced
-/+ resource "aws_eks_cluster" "eks" {
      ~ arn                       = "arn:aws:eks:us-east-1:<ACCOUNTID>:cluster/devcluster" -> (known after apply)
      ~ certificate_authority     = [
          - {
              - data = ""
            },
        ] -> (known after apply)
      ~ created_at                = "2022-03-09 05:32:22.29 +0000 UTC" -> (known after apply)
      ~ endpoint                  = "https://<HIDE>.gr7.us-east-1.eks.amazonaws.com" -> (known after apply)
      ~ id                        = "devcluster" -> (known after apply)
      ~ identity                  = [
          - {
              - oidc = [
                  - {
                      - issuer = "https://oidc.eks.us-east-1.amazonaws.com/id/<HIDE>"
                    },
                ]
            },
        ] -> (known after apply)
        name                      = "devcluster"
      ~ platform_version          = "eks.4" -> (known after apply)
      ~ status                    = "ACTIVE" -> (known after apply)
        tags                      = {
            "application" = "infrastructure"
            "email"       = "sre@domain.com"
            "environment" = "dev"
            "owner"       = "sre"
            "product"     = "infrastructure"
        }
        # (4 unchanged attributes hidden)

      ~ kubernetes_network_config {
          ~ ip_family         = "ipv4" -> (known after apply)
          ~ service_ipv4_cidr = "172.20.0.0/16" -> (known after apply)
        }


      ~ vpc_config {
          ~ cluster_security_group_id = "sg-0b77528c63a079fce" -> (known after apply)
          ~ subnet_ids                = [
              + "subnet-03598955ac92677f2",
              + "subnet-05313c98a028c4022",
                # (3 unchanged elements hidden)
            ]
          ~ vpc_id                    = "vpc-0091983ba4608d291" -> (known after apply)
            # (4 unchanged attributes hidden)
        }
        # (1 unchanged block hidden)
    }

  # module.eks_cluster.null_resource.wait_for_api must be replaced
-/+ resource "null_resource" "wait_for_api" {
      ~ id       = "1451658124690935433" -> (known after apply)
      ~ triggers = {
          - "eks_cluster_endpoint" = "https://<HIDE>.gr7.us-east-1.eks.amazonaws.com"
          - "endpoint_access"      = "false"
          - "k8s_version"          = "1.21"
        } -> (known after apply) # forces replacement
    }

Plan: 2 to add, 0 to change, 2 to destroy.

Changes to Outputs:
  ~ eks_cluster_arn  = "arn:aws:eks:us-east-1:<ACCOUNTID>:cluster/devcluster" -> (known after apply)
  ~ eks_cluster_name = "devcluster" -> (known after apply)
  ~ eks_kubeconfig   = <<-EOT
        apiVersion: v1
        clusters:
        - cluster:
            server: https://<HIDE>.gr7.us-east-1.eks.amazonaws.com
            certificate-authority-data: 
          name: kubernetes
        contexts:
        - context:
            cluster: kubernetes
            user: aws
          name: aws
        current-context: aws
        kind: Config
        preferences: {}
        users:
        - name: aws
          user:
            exec:
              apiVersion: client.authentication.k8s.io/v1alpha1
              command: aws-iam-authenticator
              args:
                - "token"
                - "-i"
                - "devcluster"
    EOT -> (known after apply)
╷
│ Warning: Resource targeting is in effect
│
│ You are creating a plan with the -target option, which means that the
│ result of this plan may not represent all of the changes requested by the
│ current configuration.
│
│ The -target option is not for routine use, and is provided only for
│ exceptional situations such as recovering from errors or mistakes, or when
│ Terraform specifically suggests to use it as part of an error message.
╵

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: ^C
  ```

**Reply by yorinasub17** [reply]
> When I attempt your method using the targeting apply, TF wants to recreate the cluster because it is tainted. After removing the data source, should I remove module.eks_cluster.aws_eks_cluster.eks from the state file and re-import the cluster?

I am pretty sure the taint is unrelated to the data source change, since the relationship is the other way around (the data source depends on the EKS cluster). In fact, I think that taint is probably the reason why the refresh is failing - a tainted resource is indicative of a partial apply and the resource is somehow in an inconsistent state.

Assuming you have verified the EKS cluster is actually functioning, I would recommend doing [an `untaint` call on the EKS cluster resource instead](https://www.terraform.io/cli/commands/untaint).

After the untaint, try to see if the data source will apply by itself and hopefully, that should get everything back in a consistent state again.

**Reply by BrianMusson** [reply]
I followed your suggestion and ran the `untaint module.eks_cluster.aws_eks_cluster.eks` . It went through successfully, however, in this output I noticed that it says "forces replacement" on subnet_ids. This is coming up now because a while back I increased the number of private-app subnets from 3 to 6, and the [default behavior of the module is to compute the number of control plane subnets based on the list of private-app subnets.](https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/0ccedb317bfcce0e2d2a19dd83cbf3caa44ec927/examples/for-production/infrastructure-live/_envcommon/services/eks-cluster.hcl#L115) I added 4 new subnets for a total of 6, but only shows 5 because us-east-1e is defined in the disallowed var.

```
          ~ subnet_ids                = [ # forces replacement
              + "subnet-03598955ac92677f2",
              + "subnet-05313c98a028c4022",
                # (3 unchanged elements hidden)
```
                
I set [num_control_plane_vpc_subnet_ids](https://github.com/gruntwork-io/terraform-aws-service-catalog/blob/0ccedb317bfcce0e2d2a19dd83cbf3caa44ec927/modules/services/eks-cluster/variables.tf#L67) to 3 so it matches what was there originally, ran `terragrunt apply -refresh=false -target data.aws_eks_cluster.cluster` and now the `module.eks_k8s_role_mapping.kubernetes_config_map.eks_to_k8s_role_mapping` object comes back with some updates, along with a bunch of other resources.
```
  # module.eks_cluster.aws_security_group.eks has been changed
  # module.eks_cluster.aws_eks_fargate_profile.control_plane_services[0] has been deleted
  # module.eks_workers["enabled"].module.managed_node_groups.aws_eks_node_group.eks_worker["mng-t3a-xlarge"] has been changed
  # module.eks_cluster.aws_eks_fargate_profile.control_plane_services[0] will be created
  # module.eks_k8s_role_mapping.kubernetes_config_map.eks_to_k8s_role_mapping will be updated in-place
  # module.eks_workers["enabled"].module.managed_node_groups.aws_eks_node_group.eks_worker["mng-t3a-xlarge"] must be replaced
  # module.eks_workers["enabled"].module.ec2_baseline_common.module.high_asg_cpu_usage_alarms.aws_cloudwatch_metric_alarm.asg_high_cpu_utilization[0] must be replaced
  # module.eks_workers["enabled"].module.ec2_baseline_common.module.high_asg_disk_usage_root_volume_alarms.aws_cloudwatch_metric_alarm.asg_high_disk_utilization[0] must be replaced
  # module.eks_workers["enabled"].module.ec2_baseline_common.module.high_asg_memory_usage_alarms.aws_cloudwatch_metric_alarm.asg_high_memory_utilization[0] must be replaced
  ``` 

Thanks for the hints and the help. That was fun!

**Reply by yorinasub17** [reply]
Ahhhh that all makes sense now! Presumably what happened was that the EKS cluster was tainted and everything in the state was invalidated due to the subnet change, which then prevented terraform from properly setting up the `kubernetes` provider.

Glad to hear you were able to make forward progress with the hints above!
